packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -326,6 +333,70 @@ function parseBaseElement(
     return element;
 }
 
+function parseLwcElement(
+    ctx: ParserCtx,
+    parse5Elm: parse5.Element,
+    parsedAttr: ParsedAttribute,
+    parse5ElmLocation: parse5.ElementLocation
+) {
+    let lwcElementParser;
+
+    switch (parse5Elm.tagName) {
+        case ctx.config.enableDynamicComponents && LwcTagName.Component:
```

Comment:
This is kind of weird to me. So if `ctx.config.enableDynamicComponents` is false, then we will just treat `lwc:component` as an unknown HTML element? I guess it's fine, though, since throwing an error here is kind of pointless since we will just have to get rid of it when the `enableDynamicComponents` flag is removed.

Reply:
Ah I see, we are warning below. OK, this makes sense then.

Reply:
Agree, this switch statement is weird. @jmsjtu IMO, the checking of `ctx.config.enableDynamicComponents` should be inside `parseLwcComponent`. `parseLwcComponent` can validate if a template has `lwc:is` without enabling `enableDynamicComponents` and warn. This way all the validation for `lwc:is` is consolidated in one routine.

Reply:
@nolanlawson @ravijayaramappa after some thought I've removed the `LWC_COMPONENT_USED_WITHOUT_OPT` warning.

The original intent of the warning was to notify component authors that `lwc:component` would be treated as an unknown HTML element if the `ctx.config.enableDynamicComponents` is not enabled.  

However, now that I think about it, this can be confusing.  I've opted instead to treat `lwc:component` the same regardless of whether `ctx.config.enableDynamicComponents` is enabled.

This way `parseLwcComponent` [checks](https://github.com/salesforce/lwc/blob/7be5c7841139d3908ab1841bb09810dec034b5bd/packages/%40lwc/template-compiler/src/parser/index.ts#LL992) to see that `lwc:component` must be used with `lwc:is` and `applyLwcIsDirective` checks to see that `ctx.config.enableDynamicComponents` is enabled to use `lwc:is`.

The [latest commit](https://github.com/salesforce/lwc/pull/3337/commits/7be5c7841139d3908ab1841bb09810dec034b5bd) has the updated logic, let me know what you think.

Reply:
@jmsjtu So what happens if `enableDynamicComponents` is `false` but the template contains `lwc:component`?

Reply:
If `enableDynamicComponents` is `false` and a template contains a `lwc:component` without `lwc:is` it will [throw an error saying you need to use `lwc:is` with `lwc:component`](https://github.com/salesforce/lwc/blob/7be5c7841139d3908ab1841bb09810dec034b5bd/packages/%40lwc/template-compiler/src/__tests__/fixtures/dynamic-components/invalid/no-config/metadata.json).

If `enableDynamicComponents` is false and have `lwc:component` + `lwc:is` you will get an [error saying `enableDynamicComponents` must be enabled](https://github.com/salesforce/lwc/blob/7be5c7841139d3908ab1841bb09810dec034b5bd/packages/%40lwc/template-compiler/src/__tests__/fixtures/dynamic-components/invalid/invalid-config/metadata.json).

In either case if `enableDynamicComponents` is `false` or `true` we treat `lwc:component` the same and only validate the config on `lwc:is`.

Reply:
IMO, in both case #1 and #2, the error should say that `enableDynamicComponents must be enabled to use <lwc:component> element`. In other words, `enableDynamicComponents` flag gates the usage of both `lwc:component` and/or `lwc:is`.

Reply:
@nolanlawson @ravijayaramappa  - I decided to move the check for `enableDynamicComponents` from [`applyLwcIsDirective`](https://github.com/salesforce/lwc/pull/3337/commits/98211cf40d4cc4a843e7d09b8d4703eafd4dfd4e#diff-7d2d3ce772f6d0a96d43e3579efb79f9c91407e26851f4500db2c8181f5f4aa4R998-R999) to [`parseLwcComponent`](https://github.com/salesforce/lwc/pull/3337/commits/98211cf40d4cc4a843e7d09b8d4703eafd4dfd4e#diff-7d2d3ce772f6d0a96d43e3579efb79f9c91407e26851f4500db2c8181f5f4aa4R361-R367).

I think this way the validation is cleaner, here are the [new unit tests](https://github.com/salesforce/lwc/blob/98211cf40d4cc4a843e7d09b8d4703eafd4dfd4e/packages/%40lwc/template-compiler/src/__tests__/fixtures/dynamic-components/invalid/invalid-config/actual.html).

Let me know if you think should be changed.  I'm going to merge this PR but it'll just go into the` jtu/dynamic-components` branch.

Sorry for the back and forth on this!